### PR TITLE
[IMP] Store the original constraint definition as a comment

### PR DESCRIPTION
**Pull request by StefanRijnhart**
Store the original constraint definition as a comment on the constraint
and use it to determine if the constraint definition was changed. This prevents
endless re-adding of constraints that are reformatted by postgresql in an
unrecognizable way (in which e.g. "CHECK (credit*debit=0)" becomes
"CHECK ((credit * debit) = 0::numeric)")

Description of the issue/feature this PR addresses:
Existing SQL constraints are not properly recognized on module update

Current behavior before PR:
An update of the account module drops and re-adds _sql_constraints such as those on move and move line balance which can take a long time on large databases.

Desired behavior after PR is merged:
If the constraint definition is not modified, the constraint is not dropped and re-added

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by StefanRijnhart** [comment]
We can start adding constraint comments in Odoo 11.0 to prevent all of them from being re-added in the migration to 12.0

**Comment by rco-odoo** [comment]
@StefanRijnhart can you give some actual numbers about how much you can gain from this?

**Comment by rco-odoo** [comment]
@StefanRijnhart thanks for the info, that's definitively worth it!

**Comment by StefanRijnhart** [comment]
I considered it but I liked the existing clean API in tools/sql.py. If you insist I can combine the methods.

**Comment by rco-odoo** [comment]
@StefanRijnhart I also like the clean API. And both `pg_constraint.conname` and `pg_description.objoid` are indexed, so `constraint_comment` is efficient... Keep it as it is.

File: odoo/tools/sql.py [hunk 1]
```diff
@@ -123,12 +123,20 @@ def constraint_definition(cr, constraintname):
...
+    return cr.fetchone()[0] if cr.rowcount else None
+
 def add_constraint(cr, tablename, constraintname, definition):
     """ Add a constraint on the given table. """
     query = 'ALTER TABLE "{}" ADD CONSTRAINT "{}" {}'.format(tablename, constraintname, definition)
     try:
         with cr.savepoint():
             cr.execute(query)
+            cr.execute(
+                'COMMENT ON CONSTRAINT "{}" ON "{}" IS %s'.format(constraintname, tablename), (definition,))
```
Comment (pedrobaeza) [inline odoo/tools/sql.py]: Isn't here the possibility of SQL injection?

Comment (StefanRijnhart) [reply odoo/tools/sql.py]: Indeed @rco-odoo you should probably use psycopg2.sql.Identifier everywhere in this file (t.i.l. AsIs is not safe either). See https://stackoverflow.com/a/42947632

Comment (rco-odoo) [reply odoo/tools/sql.py]: @pedrobaeza you could consider an SQL injection if the constraint definition was given by some unchecked program input. But that's not the case: the constraint definition is in your program itself.

File: odoo/tools/sql.py [hunk 2]
```diff
@@ -123,12 +123,20 @@ def constraint_definition(cr, constraintname):
     cr.execute("SELECT pg_get_constraintdef(oid) FROM pg_constraint WHERE conname=%s", (constraintname,))
     return cr.fetchone()[0] if cr.rowcount else None
 
+def constraint_comment(cr, constraintname):
+    """ Return the given constraint's comment. """
+    cr.execute("SELECT description FROM pg_description d JOIN pg_constraint c "
+               "ON c.oid = d.objoid AND conname = %s;", (constraintname,))
```
Comment (odony) [inline odoo/tools/sql.py]: Constraint names aren't unique, so even if the framework prefixes them with the model name, may I suggest we fully qualify them with the table?

```python
query = """
    SELECT description FROM pg_description d
                       JOIN pg_constraint c ON (c.oid = d.objoid AND conname = %s)
                       JOIN pg_class t ON (t.oid = c.conrelid AND t.relname = %s)
"""
cr.execute(query, (constraintname, tablename))
```
(requires passing tablename)

We should probably do the same in `constraint_definition()`, etc.

File: odoo/models.py [hunk 3]
```diff
@@ -2277,16 +2277,13 @@ def _add_sql_constraints(self):
...
-        def cons_text(txt):
-            return txt.lower().replace(', ',',').replace(' (','(')
-
         def process(key, definition):
             conname = '%s_%s' % (self._table, key)
             has_definition = tools.constraint_definition(cr, conname)
             if not has_definition:
                 # constraint does not exists
                 tools.add_constraint(cr, self._table, conname, definition)
-            elif cons_text(definition) != cons_text(has_definition):
```
[the remark this answers is not in the document]
Comment (StefanRijnhart) [reply odoo/models.py]: To be honest, I don't think it is worth the effort, if possible at all: discarding parens in cons_text could lead to false positives, i.e. a constraint that is not updated when a scoping bug was fixed.

As discussed above, the problem is not that the index is re-added once (in this case during a migration), but that it is re-added repeatedly when modules are upgraded that override the model with the problematic constraint.

File: odoo/tools/sql.py (hunk 1 above)
Comment (rco-odoo) [reply odoo/tools/sql.py]: @StefanRijnhart may I suggest to define your query as the one above?
```python
query1 = 'ALTER TABLE...'.format(...)
query2 = 'COMMENT ON CONSTRAINT...'.format(...)
with cr.savepoint():
    cr.execute(query1)
    cr.execute(query2, [definition])
```

Comment (StefanRijnhart) [reply odoo/tools/sql.py]: Done

**Comment by rco-odoo** [comment]
@StefanRijnhart after some discussion with @odony, I propose you to merge both functions `constraint_definition` and `constraint_comment`, in order to simplify their API and to hide the implementation detail (definition is "stored" in comment). That will make the code that uses them more understandable:
```python
current_definition = tools.constraint_definition(cr, self._table, conname)
if not current_definition:
    # constraint does not exists
    ...
elif current_definition != definition
    # constraint exists but its definition may have changed
    ...
```
Given the new convention (comment contains original definition), the function `constraint_definition` queries the constraint comment, with a fallback in case it does not exist:
```python
def constraint_definition(cr, tablename, constraintname):
     """ Return the given constraint's definition. """
    # retrieve the constraint's original definition from its comment if possible
    query = """
        SELECT COALESCE(d.description, pg_get_constraintdef(c.oid))
        FROM pg_constraint c
        JOIN pg_class t ON t.oid = c.conrelid
        LEFT JOIN pg_description d ON d.objoid = c.oid
        WHERE t.relname = %s AND c.conname = %s
    """
    cr.execute(query, (tablename, constraintname))
    return cr.fetchone()[0] if cr.rowcount else None
```
What do you think?

**Comment by mreficent** [comment]
@StefanRijnhart ?

**Comment by StefanRijnhart** [comment]
@mreficent thanks for the ping, I missed this one. @rco-odoo agreed. Pushed another commit with your code.

File: odoo/tools/sql.py [hunk 4]
```diff
@@ -118,17 +118,25 @@ def drop_not_null(cr, tablename, columnname):
...
+        LEFT JOIN pg_description d ON c.oid = d.objoid
+        WHERE t.relname = %s AND conname = %s;"""
+    cr.execute(query, (tablename, constraintname))
     return cr.fetchone()[0] if cr.rowcount else None
 
 def add_constraint(cr, tablename, constraintname, definition):
     """ Add a constraint on the given table. """
-    query = 'ALTER TABLE "{}" ADD CONSTRAINT "{}" {}'.format(tablename, constraintname, definition)
+    query1 = 'ALTER TABLE "{}" ADD CONSTRAINT "{}" {}'.format(tablename, constraintname, definition)
+    query2 = 'COMMENT ON CONSTRAINT "{}" ON "{}" IS %s'.format(constraintname, tablename)
```
Comment (rco-odoo) [inline odoo/tools/sql.py]: There is a small at the end of the function: variable `query` is undefined (in the `except` section).
Replace it by `query1`.

Comment (StefanRijnhart) [reply odoo/tools/sql.py]: Done

**Pull request merged**